src/dataworker/Dataworker.ts

```
@@ -148,18 +149,71 @@ export class Dataworker {
+          if (refunds[addressA].gt(refunds[addressB])) return -1;
+          if (refunds[addressA].lt(refunds[addressB])) return 1;
+          const sortOutput = this.sortAddresses(addressA, addressB);
+          if (sortOutput !== 0) return sortOutput;
+          else throw new Error("Unexpected matching address");
+        });
+
+        // Create leaf for { repaymentChainId, L2TokenAddress }, split leaves into sub-leaves if there are too many
+        // refunds.
+        for (let i = 0; i < sortedRefundAddresses.length; i += this.clients.configStoreClient.maxRefundsPerLeaf)
+          relayerRefundLeaves.push({
+            groupIndex: i, // Will delete this group index after using it to sort leaves for the same chain ID and
+            // L2 token address
+            leafId: 0, // Will be updated before inserting into tree when we sort all leaves.
+            chainId: Number(repaymentChainId),
+            amountToReturn: toBN(0), // TODO: Derive amountToReturn
+            l2TokenAddress,
+            refundAddresses: sortedRefundAddresses.slice(i, i + this.clients.configStoreClient.maxRefundsPerLeaf),
+            refundAmounts: sortedRefundAddresses
+              .slice(i, i + this.clients.configStoreClient.maxRefundsPerLeaf)
+              .map((address) => refunds[address]),
+          });
+      });
+    });
+
+    // Sort leaves by chain ID and then L2 token address in ascending order. Assign leaves unique, ascending ID's
+    // beginning from 0.
+    const indexedLeaves: RelayerRefundLeaf[] = [...relayerRefundLeaves]
+      .sort((leafA, leafB) => {
+        if (leafA.chainId !== leafB.chainId) {
+          return leafA.chainId - leafB.chainId;
+        } else if (this.sortAddresses(leafA.l2TokenAddress, leafB.l2TokenAddress) !== 0) {
+          return this.sortAddresses(leafA.l2TokenAddress, leafB.l2TokenAddress);
+        } else return leafA.groupIndex - leafB.groupIndex;
```

Should we throw an error if these group indices match?

---

yea good point. Do you think we should refactor this whole sorting logic into a utility function and unit test it?